Added line-up-words

diml opened the pull request:
### Brief summary of what the package does

This package provides emacs integration for the `line-up-words` tool that aligns words in a region. For instance, selecting lines 2-10 in the following code and pressing `M-x line-up-words` align the code as shown below:

```ocaml
type package =
  { name : string
  ; dir : Path.t
  ; version : string
  ; description : string
  ; archives : Path.t list Mode.Dict.t
  ; plugins : Path.t list Mode.Dict.t
  ; jsoo_runtime : string list
  ; requires : t list
  ; ppx_runtime_deps : t list
  }
```

```ocaml
type package =
  { name             : string
  ; dir              : Path.t
  ; version          : string
  ; description      : string
  ; archives         : Path.t list Mode.Dict.t
  ; plugins          : Path.t list Mode.Dict.t
  ; jsoo_runtime     : string list
  ; requires         : t list
  ; ppx_runtime_deps : t list
  }
```

### Direct link to the package repository

https://github.com/janestreet/line-up-words

### Your association with the package

This package is developed by [Jane Street](https://www.janestreet.com/). I'm a developer at Jane Street and I'm a member the team that takes care of publishing our open source packages on github.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses). 
- [x] I've read [CONTRIBUTING.md](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.md)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.md](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.md)

diml commented:
/cc @mkoconnor

purcell commented:
Thanks! A few quick notes:

- You should omit the `:files` spec from the recipe, since the default is fine here.
- Does the underlying tool do anything that's fundamentally smarter than Emacs' own `align-regexp` function, which has a wide range of general and language-specific rules (see `align-rules-list`? (Similarly, in the example above, `M-x align-regexp` then `:` appears to give the same result.)
- The executable name should probably be a `defcustom`, associated with a new `defgroup`.
- The `Commentary` section refers people to the github README, so package users will see nothing about the package on its MELPA page or in the `list-packages` menu. Better to inline some docs there, in the self-documenting Emacs spirit. Also, there's no information provided about how to install the associated tool, which will deter people from using this.

Hope that helps!

diml commented:
Thanks for your comments! I didn't know about `align-regexp`, I'll ask to get more details about it

ninrod commented:
Hi @diml. Also see [evil-lion](https://github.com/edkolev/evil-lion). has the same idea, but applies to the [evil](https://github.com/emacs-evil/evil) ecosystem.

diml commented:
Hi @ninrod, thanks I didn't know about evil-lion either.

/cc @yminsky, who just told me that he has been using `align-regexp` when programming outside of work and he doesn't find it as good as `line-up-words`

yminsky commented:
align-regexp only aligns a single regexp, but this one can handle more complicated cases. For example, it will turn this:

``` ocaml
val foo : int -> float option
val snooble : string -> int option
```

into this.

``` ocaml
val foo     : int    -> float option
val snooble : string -> int   option
```

I believe it also handles some special cases in OCaml, like this one:

``` ocaml
type t =
  { foo : int -> foop 
  ; peeble : bleep -> bar
  (* this is some : stuff *)
  ; groupple : grimpgrimp -> gromp
  }                          
```

``` ocaml
type t =
  { foo      : int        -> foop 
  ; peeble   : bleep      -> bar  
  (* this is some : stuff *)
  ; groupple : grimpgrimp -> gromp
  }                          
```

Which I think involves some sensitivity to OCaml comments. @mkoconnor would know more about what it does.

mkoconnor commented:
Yep, that’s basically it. The other difference is that line-up-words will attempt to not change spacing inside quoted strings (although it only looks at the region it’s aligning, so in theory it could be wrong about what’s actually inside a quoted string, although in practice it seems to work well)

On Thu, Feb 15, 2018 at 11:48 AM, Yaron Minsky <notifications@github.com> wrote:

> align-regexp only aligns a single regexp, but this one can handle more complicated cases. For example, it will turn this:
>
> val
>
> foo
>
> :
>
> int
>
> ->
>
> float
>
> option
>
> val
>
> snooble
>
> :
>
> string
>
> ->
>
> int
>
> option
>
> into this.
>
> val
>
> foo
>
> :
>
> int
>
> ->
>
> float
>
> option
>
> val
>
> snooble
>
> :
>
> string
>
> ->
>
> int
>
> option
>
> I believe it also handles some special cases in OCaml, like this one:
>
> type
>
> t
>
> =
>   {
>
> foo
>
> :
>
> int
>
> ->
>
> foop
>
> ;
>
> peeble
>
> :
>
> bleep
>
> ->
>
> bar
>
> (*
>
> this is some : stuff
>
> *)
>
> ;
>
> groupple
>
> :
>
> grimpgrimp
>
> ->
>
> gromp
>
> }
>
> type
>
> t
>
> =
>   {
>
> foo
>
> :
>
> int
>
> ->
>
> foop
>
> ;
>
> peeble
>
> :
>
> bleep
>
> ->
>
> bar
>
> (*
>
> this is some : stuff
>
> *)
>
> ;
>
> groupple
>
> :
>
> grimpgrimp
>
> ->
>
> gromp
>
> }
>
> Which I think involves some sensitivity to OCaml comments. [@mkoconnor](https://github.com/mkoconnor) would know more about what it does.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/melpa/melpa/pull/5305#issuecomment-365988569), or [mute the thread](https://github.com/notifications/unsubscribe-auth/AA0WHpd-8BW8YuJkbZDHsv2mvJ7OvV2pks5tVF_XgaJpZM4SD2U3).
>
> {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/melpa/melpa","title":"melpa/melpa","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/melpa/melpa"}},"updates":{"snippets":[{"icon":"PERSON","message":"@yminsky in #5305: align-regexp only aligns a single regexp, but this one can handle more complicated cases. For example, it will turn this: ``` ocaml val foo : int -> float option val snooble : string -> int option ``` into this. ``` ocaml val foo : int -> float option val snooble : string -> int option ``` I believe it also handles some special cases in OCaml, like this one: ``` ocaml type t = { foo : int -> foop ; peeble : bleep -> bar (* this is some : stuff *) ; groupple : grimpgrimp -> gromp } ``` ``` ocaml type t = { foo : int -> foop ; peeble : bleep -> bar (* this is some : stuff *) ; groupple : grimpgrimp -> gromp } ``` Which I think involves some sensitivity to OCaml comments. @mkoconnor would know more about what it does."}],"action":{"name":"View Pull Request","url":"https://github.com/melpa/melpa/pull/5305#issuecomment-365988569"}}}

purcell commented:
Thanks all. Yeah, `align-regexp` alone certainly isn't comparable to `line-up-words`, but the `align` machinery in Emacs should be (if tweaked), so if it's not doing a nice job of aligning ocaml code now then an ideal way help Emacs users in the long term would be to create `align` rules, and either publish them separately, include them in `caml-mode` or even submit them back into the default Emacs ruleset.

That said, I understand the value of reformatting executables, and there's no harm in adding this package here, subject to covering the couple of little points in my previous comment.

edkolev commented:
@diml Purcell mentioned this above, but I think it's worth repeating - this package has no installations instructions for the binary, I doubt many users will be able to use it.

diml commented:
Sure, I'm going update the package and PR to take into account the review comments

diml commented:
Ok, I removed the `:files`, added some doc with installation instructions and added a `defcustom` with a group.

I agree that it'd be nicer to not require an external tool, in particular the installation part might look complicated for non OCaml users. It's just that we are more fluent in OCaml than Emacs lisp :)

purcell commented:
Great - merged, thanks!

diml commented:
Thanks :)
